### PR TITLE
Port LIT tests to use runtime functions

### LINKED ISSUE #60: Add LIT test that uses the runtime
**Description:**
Add LIT tests to cover the functionality added by the runtime

**Suggested Labels:**
Priority/High

**Suggested Assignees:**
shubhamnarlawar

### LINKED ISSUE #65: LIT test floatTypecast fails to compile
**Description:**
The LIT test [floatTypecast.bal](https://github.com/ballerina-platform/nballerina/blob/master/test/casting/floatTypecast.bal.skip) fails generate a .ll throwing the following error.
```
./nballerinacc temp-dump
nballerinacc: /home/ruvinda/projects/nballerina/compiler/BIRReader.cpp:120: float ConstantPoolSet::getFloatCp(uint32_t): Assertion `poolEntry->getTag() != ConstantPoolEntry::tagEnum::TAG_ENUM_CP_ENTRY_FLOAT' failed.
Aborted (core dumped)
```

**Suggested Labels:**
Type/Bug Priority/High 

**Suggested Assignees:**
shubhamnarlawar

### LINKED ISSUE #153: Add rust function print_str() to print ballerina string
**Description:**
Add support for printing bal string using rust function.

### PR DESCRIPTION
**Pull request by shubhamnarlawar**
Port existing LIT unit tests to use runtime functions.

Resolves #60
Resolves #65
Resolves #153

### CODE REVIEW

**Comment by shubhamnarlawar** [comment]
Is this PR ready to merge in master?

File: compiler/BIRReader.cpp [hunk 1]
```diff
@@ -667,6 +667,7 @@ Function *BIRReader::readFunction(Package *package) {
     Function *birFunction = new Function(package, functionName, constantPool->getStringCp(workdernameCpIndex), flags,
                                          constantPool->getInvokableType(typeCpIndex));
     birFunction->setLocation(location);
+    birFunction->setFlags(flags);
```
Comment (ruvi-d) [inline compiler/BIRReader.cpp:670]: flags are being passed via the constructor in line 667. No need to set it again.
Just update the constructor to initialize flags.

File: compiler/codegen/Function.cpp [hunk 2]
```diff
@@ -112,6 +113,7 @@ void Function::insertLocalVar(Variable *var) {
     localVars.insert(std::pair<std::string, Variable *>(var->getName(), var));
 }
 void Function::setReturnVar(Variable *var) { returnVar = var; }
+void Function::setFlags(unsigned int f) { flags = f; }
```
Comment (ruvi-d) [inline compiler/codegen/Function.cpp:116]: Flags setter not required

File: compiler/include/Function.h [hunk 3]
```diff
@@ -38,11 +38,16 @@ class RestParam;
 class Type;
 class Package;
 
+// Function flags
+static constexpr unsigned int PUBLIC = 1;
+static constexpr unsigned int NATIVE = PUBLIC << 1;
+
```
Comment (ruvi-d) [inline compiler/include/Function.h:44]: I'd suggest that you make these constants private to the Function class and expose it to Package class via a `isNatveFunction()`. You can do the `(function->getFlags() & NATIVE) == NATIVE` logic inside this function.

File: compiler/codegen/Package.cpp [hunk 4]
```diff
@@ -151,6 +151,8 @@ void Package::translate(LLVMModuleRef &modRef) {
 
     // iterating over each function translate the function body.
     for (const auto &function : functions) {
+        if ((function->getFlags() & NATIVE) == NATIVE)
+            continue;
```
Comment (ruvi-d) [inline compiler/codegen/Package.cpp:155]: See other comment about using a `isNatveFunction()`

**Review by ruvi-d — CHANGES_REQUESTED** [review]

**Comment by ruvi-d** [comment]
BTW, what happen to the check if return type is VOID from your previous PR?

**Comment by shubhamnarlawar** [comment]
> BTW, what happen to the check if return type is VOID from your previous PR?

There is one comment by you on PR #159 -

`Since #94 we handling returning NIL type via pointer. So there is no need to check for void return types, because we shouldn't be generating any functions that have void types.`

Now, there are no void type function. Hence, that check from other PR is not required.

File: compiler/codegen/Package.cpp (hunk 4 above)
Comment (shubhamnarlawar) [reply compiler/codegen/Package.cpp]: Done

File: compiler/include/Function.h (hunk 3 above)
Comment (shubhamnarlawar) [reply compiler/include/Function.h]: Done

File: compiler/codegen/Function.cpp (hunk 2 above)
Comment (shubhamnarlawar) [reply compiler/codegen/Function.cpp]: Removed

File: compiler/BIRReader.cpp (hunk 1 above)
Comment (shubhamnarlawar) [reply compiler/BIRReader.cpp]: Done.

**Comment by shubhamnarlawar** [comment]
@ruvi-d - I have addressed and pushed changes of above suggestions. Is it ready to merge now?

File: compiler/codegen/Function.cpp [hunk 5]
```diff
@@ -29,8 +29,8 @@ namespace nballerina {
 
 Function::Function(Package *_parentPackage, std::string name, std::string workerName, [[maybe_unused]] int flags,
```
Comment (ruvi-d) [inline compiler/codegen/Function.cpp:30]: Remove the `[[maybe_unused]]` in front of `int flags`

**Review by ruvi-d — APPROVED** [review]
Just one more minor thing. Good to merge afterwards.

File: compiler/codegen/Function.cpp (hunk 5 above)
Comment (shubhamnarlawar) [reply compiler/codegen/Function.cpp]: Done

**Comment by shubhamnarlawar** [comment]
Is it fine to merge now?

**Review by Kishanthan — APPROVED** [review]

**Comment by ruvi-d** [comment]
Pending tasks: #170, #171

**Comment by ayonam** [comment]
@shubhamnarlawar  The issue that you are facing is because of the compiler not finding the ballerina_rt library in the directory that you gave as the library search path.  In the command below, you provide only one path (../../../runtime/target/release) as the library search path.  

              clang -O0 -o $filename.out $filename-bir-dump.ll -L../../../runtime/target/release -lballerina_rt 2>err.log

Since it is a relative path, it may be possible that the directory from where you are running the failing test case, this path does not exist as a relative path and hence it is unable to find the library file  libballerina_rt.so.  So you should investigate one failing test case and check the directory from where it is invoked.  Then see if it the library file is reachable from there using this path.  If not, then you need to add the relevant path to the library search list using another -L option on the command line apart from the one that you already have.

**Pull request merged**